### DISCUSSION TITLE
Led-Ring and Flowdeck  -  Firmware Version Conflict

### DISCUSSION THREAD

**Discussion by magalhaesw** [bitcraze/discussions#795] [Q&A] [answered]
Dear All,
Good Evening.

Id like to ask for help.
I am using the CF2 with a flowdeck on the bottom and a led-ring on the top.

I followed the pin-layout LED available at : https://www.bitcraze.io/products/led-ring-deck/
and shown in the figure below,
![Led_Ring_Pinout](https://github.com/bitcraze/discussions/assets/53016613/55991cf6-b287-4344-ac4d-9675a924c9bd)

The CF2 works well with firmware version **2019.02.0**1, 
But it does not work with later version.

For 2022 firmware version, I get the following error:
```
> DECK_INFO: ERROR: Driver Gpio usage conflicts with a previously enumerated deck driver. No decks will be initialized!
> DECK_CORE: 0 deck(s) found
```

Please,
What can I do to solve this issue ?
Is the pin-layout for the led-ring on the top different for the later version ?

Thanks for your help.

Best,
    Jose

**Comment by krichardsson** [comment, marked as answer]
Hi,

Each deck driver has a list of the GPIO pins that is used by the deck, and if the firmware detects that two decks use the same GPIO you will get the error message you see. Both the Flow deck and the LED ring uses GPIO 3, so it is correct that this happens. 

In your case it looks like you did not wire GPIO 3 on your LED-ring conversion (it is used for the front LEDs) which means that there will not be a conflict between the decks, which is good :-) What you need to do is to update the LED-ring deck driver to let the system know that GPIO3 is not used by the deck. Simply remove `DECK_USING_IO_3' from [the deck driver](https://github.com/bitcraze/crazyflie-firmware/blob/41a9c97569e565e927a0fe3b8f032dbcdb03aa1b/src/deck/drivers/src/ledring12.c#L1250), recompile and flash. We will update the instructions on the web as well.

The reason that it works in the 2019.02 firmware is simply that there was a bug and the GPIO pins were not registered properly in the deck drivers.

It seems as as the [Deck pin allocation matrix on our web](https://www.bitcraze.io/documentation/system/platform/cf2-expansiondecks/#deck-pin-allocation) has an error for GPIO 3 on the flow deck (we will look into this) as it is stated as an optional usage. [The schematics](https://www.bitcraze.io/documentation/hardware/flow_deck_2/flow-deck-v2-reva.pdf) on the other hand shows that it is used.

**Reply by magalhaesw** [reply]
Hi !
Thanks for your explanation. 
I will remove "DECK_USING_IO_3" from the firmware code.

So this part of the code will look like:

**From:**
`.usedGpio = DECK_USING_IO_2 | DECK_USING_IO_3,
`

**To:**
`.usedGpio = DECK_USING_IO_2,`


Thanks,
     Jose
